builtin/credential/appgroup/path_app.go

```diff
@@ -0,0 +1,931 @@
+	b.appLock.Lock()
+	defer b.appLock.Unlock()
+	if entry, err := logical.StorageEntryJSON("app/"+strings.ToLower(appName), app); err != nil {
+		return err
+	} else {
+		return s.Put(entry)
+	}
+}
+
+// appEntry grabs the read lock and fetches the options of an App from the storage
+func (b *backend) appEntry(s logical.Storage, appName string) (*appStorageEntry, error) {
+	if appName == "" {
+		return nil, fmt.Errorf("missing app_name")
+	}
+
+	var result appStorageEntry
+
+	b.appLock.RLock()
+	defer b.appLock.RUnlock()
+
+	if entry, err := s.Get("app/" + strings.ToLower(appName)); err != nil {
+		return nil, err
+	} else if entry == nil {
+		return nil, nil
+	} else if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+// pathAppCreateUpdate registers a new App with the backend or updates the options
+// of an existing App
+func (b *backend) pathAppCreateUpdate(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	appName := data.Get("app_name").(string)
+	if appName == "" {
+		return logical.ErrorResponse("missing app_name"), nil
+	}
+
+	// Fetch or create an entry for the app
+	app, err := b.appEntry(req.Storage, appName)
+	if err != nil {
+		return nil, err
+	}
+	// Create a new entry object if this is a CreateOperation
+	if app == nil {
+		app = &appStorageEntry{}
+	}
+
+	if policiesRaw, ok := data.GetOk("policies"); ok {
+		app.Policies = policyutil.ParsePolicies(policiesRaw.(string))
+	} else if req.Operation == logical.CreateOperation {
+		app.Policies = policyutil.ParsePolicies(data.Get("policies").(string))
+	}
+
+	if numUsesRaw, ok := data.GetOk("num_uses"); ok {
+		app.NumUses = numUsesRaw.(int)
+	} else if req.Operation == logical.CreateOperation {
+		app.NumUses = data.Get("num_uses").(int)
+	}
+	if app.NumUses < 0 {
+		return logical.ErrorResponse("num_uses cannot be negative"), nil
+	}
+
+	if userIDTTLRaw, ok := data.GetOk("userid_ttl"); ok {
+		app.UserIDTTL = time.Second * time.Duration(userIDTTLRaw.(int))
+	} else if req.Operation == logical.CreateOperation {
+		app.UserIDTTL = time.Second * time.Duration(data.Get("userid_ttl").(int))
+	}
+
+	if tokenTTLRaw, ok := data.GetOk("token_ttl"); ok {
+		app.TokenTTL = time.Second * time.Duration(tokenTTLRaw.(int))
+	} else if req.Operation == logical.CreateOperation {
+		app.TokenTTL = time.Second * time.Duration(data.Get("token_ttl").(int))
+	}
+
+	if tokenMaxTTLRaw, ok := data.GetOk("token_max_ttl"); ok {
+		app.TokenMaxTTL = time.Second * time.Duration(tokenMaxTTLRaw.(int))
+	} else if req.Operation == logical.CreateOperation {
+		app.TokenMaxTTL = time.Second * time.Duration(data.Get("token_max_ttl").(int))
+	}
+
+	// Check that the TokenMaxTTL value provided is less than the TokenMaxTTL.
+	// Sanitizing the TTL and MaxTTL is not required now and can be performed
+	// at credential issue time.
+	if app.TokenMaxTTL > time.Duration(0) && app.TokenTTL > app.TokenMaxTTL {
```

Comment:
Something I'd consider optional is if you want, you could check whether the TTLs are bigger than the sysview values and attach warnings to the response if so.